src/transforms/reduce/mod.rs

```diff
@@ -264,7 +269,9 @@ impl ReduceState {
             }
         }
         self.events += 1;
-        self.stale_since = Instant::now();
+        if !force_periodic_flush {
```

Comment:
I think we'll want to set `stale_since` here if `expire_after` has passed. Otherwise, we'll be flushing every single new event immediately after we've flushed the first batch of discriminants. In other words, we want to reset back to the start of the reducing state.

Reply:
Actually, after thinking about this some, I think we can do better. `force_periodic_flush` kind of implies that forcing a flush is a feature or an action that we seek out to do with this function, but this behavior should be implicit. Here's an alternative, perhaps more clear way to do this:

What do you think about introducing a new `Instant`, say `self.last_flushed_at`, which we'll use for tracking the time since the last flush? This would allow us to keep `add_event` intact, only changing these `flush_into` lines: https://github.com/vectordotdev/vector/blob/master/src/transforms/reduce/mod.rs#L330

To this:
```
            if t.stale_since.elapsed() >= self.expire_after {
                flush_discriminants.push(k.clone());
            } else if t.last_flushed_at.elapsed() >= self.expire_after {
                flush_discriminants.push(k.clone());
                t.last_flushed_at = Instant::now();
            }
```

That should be the correct behavior around accounting for subsequent batches of discriminants, which the proposed solution doesn't account for.

Reply:
Thanks @davidhuie-dd. I took your suggestion, and it looks much better now! I'l try to add unit test myself, however I could use some help with that :-).

Reply:
The implementation looks good! Regarding the unit test, we should be able to add VRL that demonstrates the issue, like this: https://github.com/vectordotdev/vector/blob/master/src/transforms/reduce/mod.rs#L819

I'm not 100% sure how we can make time pass, but there's probably an example in the code that we can use.

Reply:
@davidhuie-dd , I run `make test SCOPE=reduce` and it takes ~20 minutes to build and run, even after a small change to the test code. It takes 9 seconds on re-run, but then on every change it builds absolutely everything. Is there a faster way to iteratively modify. build and run tests?

Reply:
Hey @tomers ! Yeah there are some tips here for iterating more quickly: https://github.com/vectordotdev/vector/blob/master/docs/DEVELOPING.md#tips-and-tricks

In your case, in particular, I think it'll go a lot faster if you do `cargo test --no-default-features --features transforms-reduce reduce` which will only build that component.